File: backend/app/joysafeter_shared/ids.py
"""Typed entity identifiers.

Single source of truth for public-facing prefixed IDs. Each entity has a
subclass carrying the entity kind in its type (static safety) and value
(serialization/equality). Physical storage remains a bare UUID.
"""

from __future__ import annotations

import uuid
from typing import Any, ClassVar, Self

from uuid_utils import uuid7


class EntityId:
    prefix: ClassVar[str]
    __slots__ = ("_uuid",)

    def __init__(self, value: uuid.UUID | "EntityId") -> None:
        self._uuid = self._coerce(value)

    @classmethod
    def _coerce(cls, value: Any) -> uuid.UUID:
        if isinstance(value, EntityId):
            if type(value) is not cls:
                raise TypeError(f"cannot build {cls.__name__} from {type(value).__name__}")
            return value.uuid
        if isinstance(value, uuid.UUID):
            return value
        raise TypeError(f"cannot build {cls.__name__} from {type(value).__name__}")

    @classmethod
    def new(cls) -> Self:
        return cls.from_uuid(uuid.UUID(str(uuid7())))

    @classmethod
    def from_uuid(cls, value: uuid.UUID) -> Self:
        return cls(value)

    @classmethod
    def from_public(cls, value: str) -> Self:
        if not isinstance(value, str) or not value.startswith(cls.prefix):
            raise ValueError(f"expected {cls.prefix} prefix")
        raw_uuid = value[len(cls.prefix) :]
        try:
            parsed_uuid = uuid.UUID(raw_uuid)
        except ValueError as exc:
            raise ValueError(f"expected canonical {cls.prefix} entity ID") from exc
        if str(parsed_uuid) != raw_uuid:
            raise ValueError(f"expected canonical {cls.prefix} entity ID")
        return cls.from_uuid(parsed_uuid)

    @property
    def uuid(self) -> uuid.UUID:
        return self._uuid

    def __str__(self) -> str:
        return f"{self.prefix}{self._uuid}"

    def __repr__(self) -> str:
        return f"{type(self).__name__}({self._uuid})"

    def __eq__(self, other: object) -> bool:
        if type(self) is not type(other):
            return False
        assert isinstance(other, EntityId)
        return self._uuid == other._uuid

    def __lt__(self, other: object) -> bool:
        # SQLAlchemy's unit of work sorts pending objects by primary key, so a
        # PK-typed EntityId must be orderable. Order by the underlying uuid within
        # the same entity type; cross-type ordering is undefined (never mixed).
        if type(self) is not type(other):
            return NotImplemented
        assert isinstance(other, EntityId)
        return self._uuid < other._uuid

    def __hash__(self) -> int:
        return hash((type(self), self._uuid))

    @classmethod
    def __get_pydantic_core_schema__(cls, source: Any, handler: Any) -> Any:
        from pydantic_core import core_schema

        def validate(value: Any) -> "EntityId":
            if isinstance(value, cls):
                return value
            try:
                if isinstance(value, str):
                    return cls.from_public(value)
                raise TypeError(f"cannot validate {type(value).__name__} as {cls.__name__}")
            except (ValueError, TypeError):
                raise ValueError(f"__entity_id__:{cls.__name__}")  # marker for the handler

        return core_schema.no_info_plain_validator_function(
            validate,
            serialization=core_schema.plain_serializer_function_ser_schema(
                str, return_schema=core_schema.str_schema(), when_used="json-unless-none"
            ),
        )


class AgentId(EntityId):
    prefix = "agent_"


class AgentVersionId(EntityId):
    prefix = "agentver_"


class ApiKeyId(EntityId):
    prefix = "apikey_"


class UserId(EntityId):
    prefix = "user_"


class OrganizationId(EntityId):
    prefix = "org_"


class OrganizationMemberId(EntityId):
    prefix = "orgmem_"


class ProjectId(EntityId):
    prefix = "proj_"


class ProjectMemberId(EntityId):
    prefix = "projmem_"


class OAuthAccountId(EntityId):
    prefix = "oauthacct_"


class AuthSessionId(EntityId):
    prefix = "authsess_"


class SessionId(EntityId):
    prefix = "sess_"


class TaskId(EntityId):
    prefix = "task_"


class EnvironmentId(EntityId):
    prefix = "env_"


class TriggerId(EntityId):
    prefix = "trig_"


class MemoryStoreId(EntityId):
    prefix = "memstore_"


class MemoryId(EntityId):
    prefix = "mem_"


class MemoryVersionId(EntityId):
    prefix = "memver_"


class SandboxId(EntityId):
    prefix = "sbx_"


class CredentialId(EntityId):
    prefix = "cred_"


class CredentialGroupId(EntityId):
    prefix = "credgrp_"


class SkillId(EntityId):
    prefix = "skill_"


class SkillFileId(EntityId):
    prefix = "sklfile_"


class SkillSecurityScanId(EntityId):
    prefix = "sklscan_"


class SkillVersionId(EntityId):
    prefix = "sklver_"


class SkillVersionFileId(EntityId):
    prefix = "sklvfile_"


class SkillUsageId(EntityId):
    prefix = "skluse_"


class EventId(EntityId):
    prefix = "evt_"


class FileId(EntityId):
    prefix = "file_"


class SessionResourceId(EntityId):
    prefix = "sesrsc_"


class StorageVolumeId(EntityId):
    prefix = "vol_"


class StorageGrantId(EntityId):
    prefix = "stgrant_"


class StorageMountAuditId(EntityId):
    prefix = "staudit_"


class CredentialAccessAuditId(EntityId):
    prefix = "credaudit_"


class SecurityAuditId(EntityId):
    prefix = "secaudit_"


class SandboxNetworkPolicyId(EntityId):
    prefix = "sbxnetpol_"


REGISTERED_ENTITY_ID_PREFIXES: tuple[str, ...] = tuple(id_type.prefix for id_type in EntityId.__subclasses__())


def registered_entity_id_prefix(value: str) -> str | None:
    return next(
        (prefix for prefix in REGISTERED_ENTITY_ID_PREFIXES if value.startswith(prefix)),
        None,
    )


def as_uuid(value: EntityId) -> uuid.UUID:
    """Return the bare UUID for a typed entity or native UUID.

    Physical-boundary helpers (advisory-lock keys, cross-language Redis channel
    names) need the raw UUID. String parsing belongs at an explicit public or
    persistence boundary and must not leak through this adapter.
    """
    if not isinstance(value, EntityId):
        raise TypeError(f"cannot unwrap {type(value).__name__} as UUID")
    return value.uuid

File: backend/app/joysafeter_orchestrator_rs/src/ids.rs
use std::fmt;
use std::str::FromStr;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, thiserror::Error)]
pub enum EntityIdParseError {
    #[error("expected {0} prefix")]
    MissingPrefix(&'static str),
    #[error(transparent)]
    InvalidUuid(#[from] uuid::Error),
}

macro_rules! entity_id {
    ($name:ident, $prefix:literal) => {
        #[derive(
            Debug,
            Clone,
            Copy,
            PartialEq,
            Eq,
            PartialOrd,
            Ord,
            Hash,
            Serialize,
            Deserialize,
            sqlx::Type,
        )]
        #[repr(transparent)]
        #[serde(transparent)]
        #[sqlx(transparent)]
        pub struct $name(Uuid);

        impl $name {
            pub const PREFIX: &'static str = $prefix;

            pub const fn from_uuid(value: Uuid) -> Self {
                Self(value)
            }

            pub const fn as_uuid(self) -> Uuid {
                self.0
            }

            pub fn from_public(value: &str) -> Result<Self, EntityIdParseError> {
                let raw = value
                    .strip_prefix(Self::PREFIX)
                    .ok_or(EntityIdParseError::MissingPrefix(Self::PREFIX))?;
                Uuid::parse_str(raw).map(Self).map_err(Into::into)
            }

            pub fn to_public(self) -> String {
                format!("{}{}", Self::PREFIX, self.0)
            }
        }

        impl From<Uuid> for $name {
            fn from(value: Uuid) -> Self {
                Self::from_uuid(value)
            }
        }

        impl From<$name> for Uuid {
            fn from(value: $name) -> Self {
                value.as_uuid()
            }
        }

        impl FromStr for $name {
            type Err = EntityIdParseError;

            fn from_str(value: &str) -> Result<Self, Self::Err> {
                Self::from_public(value)
            }
        }

        impl fmt::Display for $name {
            fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
                write!(formatter, "{}{}", Self::PREFIX, self.0)
            }
        }
    };
}

entity_id!(AgentId, "agent_");
entity_id!(SessionId, "sess_");
entity_id!(TaskId, "task_");
entity_id!(EnvironmentId, "env_");
entity_id!(VaultId, "vault_");
entity_id!(CredentialId, "cred_");
entity_id!(SandboxId, "sbx_");
entity_id!(MemoryStoreId, "memstore_");
entity_id!(MemoryId, "mem_");
entity_id!(MemoryVersionId, "memver_");
entity_id!(SkillId, "skill_");
entity_id!(SkillFileId, "sklfile_");
entity_id!(SkillSecurityScanId, "sklscan_");
entity_id!(SkillVersionId, "sklver_");
entity_id!(SkillVersionFileId, "sklvfile_");
entity_id!(SkillUsageId, "skluse_");
entity_id!(FileId, "file_");
entity_id!(SessionResourceId, "sesrsc_");
entity_id!(EventId, "evt_");

#[cfg(test)]
mod tests {
    use super::{
        AgentId, CredentialId, EnvironmentId, EventId, FileId, MemoryId, MemoryStoreId,
        MemoryVersionId, SandboxId, SessionId, SessionResourceId, SkillFileId, SkillId,
        SkillSecurityScanId, SkillUsageId, SkillVersionFileId, SkillVersionId, TaskId, VaultId,
    };
    use uuid::Uuid;

    #[test]
    fn public_round_trip_preserves_entity_prefix() {
        let uuid = Uuid::now_v7();

        let agent_id = AgentId::from_uuid(uuid);
        let session_id = SessionId::from_uuid(uuid);
        let task_id = TaskId::from_uuid(uuid);
        let environment_id = EnvironmentId::from_uuid(uuid);
        let vault_id = VaultId::from_uuid(uuid);
        let credential_id = CredentialId::from_uuid(uuid);
        let sandbox_id = SandboxId::from_uuid(uuid);
        let store_id = MemoryStoreId::from_uuid(uuid);
        let memory_id = MemoryId::from_uuid(uuid);
        let version_id = MemoryVersionId::from_uuid(uuid);
        let skill_id = SkillId::from_uuid(uuid);
        let skill_file_id = SkillFileId::from_uuid(uuid);
        let skill_scan_id = SkillSecurityScanId::from_uuid(uuid);
        let skill_version_id = SkillVersionId::from_uuid(uuid);
        let skill_version_file_id = SkillVersionFileId::from_uuid(uuid);
        let skill_usage_id = SkillUsageId::from_uuid(uuid);
        let file_id = FileId::from_uuid(uuid);
        let session_resource_id = SessionResourceId::from_uuid(uuid);
        let event_id = EventId::from_uuid(uuid);

        assert_eq!(
            AgentId::from_public(&agent_id.to_public()).unwrap(),
            agent_id
        );
        assert_eq!(
            SessionId::from_public(&session_id.to_public()).unwrap(),
            session_id
        );
        assert_eq!(TaskId::from_public(&task_id.to_public()).unwrap(), task_id);
        assert_eq!(
            EnvironmentId::from_public(&environment_id.to_public()).unwrap(),
            environment_id
        );
        assert_eq!(
            VaultId::from_public(&vault_id.to_public()).unwrap(),
            vault_id
        );
        assert_eq!(
            CredentialId::from_public(&credential_id.to_public()).unwrap(),
            credential_id
        );
        assert_eq!(
            SandboxId::from_public(&sandbox_id.to_public()).unwrap(),
            sandbox_id
        );
        assert_eq!(
            MemoryStoreId::from_public(&store_id.to_public()).unwrap(),
            store_id
        );
        assert_eq!(
            MemoryId::from_public(&memory_id.to_public()).unwrap(),
            memory_id
        );
        assert_eq!(
            MemoryVersionId::from_public(&version_id.to_public()).unwrap(),
            version_id
        );
        assert_eq!(
            SkillId::from_public(&skill_id.to_public()).unwrap(),
            skill_id
        );
        assert_eq!(
            SkillFileId::from_public(&skill_file_id.to_public()).unwrap(),
            skill_file_id
        );
        assert_eq!(
            SkillSecurityScanId::from_public(&skill_scan_id.to_public()).unwrap(),
            skill_scan_id
        );
        assert_eq!(
            SkillVersionId::from_public(&skill_version_id.to_public()).unwrap(),
            skill_version_id
        );
        assert_eq!(
            SkillVersionFileId::from_public(&skill_version_file_id.to_public()).unwrap(),
            skill_version_file_id
        );
        assert_eq!(
            SkillUsageId::from_public(&skill_usage_id.to_public()).unwrap(),
            skill_usage_id
        );
        assert_eq!(FileId::from_public(&file_id.to_public()).unwrap(), file_id);
        assert_eq!(
            SessionResourceId::from_public(&session_resource_id.to_public()).unwrap(),
            session_resource_id
        );
        assert_eq!(
            EventId::from_public(&event_id.to_public()).unwrap(),
            event_id
        );
        assert_ne!(agent_id.to_public(), session_id.to_public());
    }

    #[test]
    fn public_parser_rejects_bare_and_cross_entity_ids() {
        let uuid = Uuid::now_v7();

        assert!(AgentId::from_public(&uuid.to_string()).is_err());
        assert!(AgentId::from_public(&SessionId::from_uuid(uuid).to_public()).is_err());
        assert!(EnvironmentId::from_public(&uuid.to_string()).is_err());
        assert!(VaultId::from_public(&uuid.to_string()).is_err());
        assert!(CredentialId::from_public(&uuid.to_string()).is_err());
    }

    #[test]
    fn storage_boundary_uses_bare_uuid() {
        let uuid = Uuid::now_v7();
        let agent_id = AgentId::from_uuid(uuid);

        assert_eq!(agent_id.as_uuid(), uuid);
        assert_eq!(Uuid::from(agent_id), uuid);
        assert_eq!(serde_json::to_value(agent_id).unwrap(), uuid.to_string());
    }
}
